### PR TITLE
Reduce builtin endpoints memory consumption <master> [7036]

### PR DESCRIPTION
**Pull request by IkerLuengo**
This is a port of #918 and #1009 from 1.9.x

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1510/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1282/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1508/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1511/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1283/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1509/)

File: src/cpp/dds/core/policy/QosPolicies.cpp [hunk 1]
```diff
@@ -593,14 +612,13 @@ bool TransportPriorityQosPolicy::readFromCDRMessage(
...
 
 bool DataRepresentationQosPolicy::addToCDRMessage(
-        CDRMessage_t* msg)
+        CDRMessage_t* msg) const
 {
     bool valid = CDRMessage::addUInt16(msg, this->Pid);
-    this->length = 4;
-    this->length += static_cast<uint16_t>(m_value.size() * sizeof(uint16_t));
-    valid &= CDRMessage::addUInt16(msg, this->length);
+    uint16_t len = 4 + static_cast<uint16_t>(m_value.size() * sizeof(uint16_t));
```
Comment (MiguelCompany) [inline src/cpp/dds/core/policy/QosPolicies.cpp]: Should account for alignment (see line 625 below). We should also add a `cdr_serialized_size` override for this class.

File: src/cpp/rtps/builtin/data/ReaderProxyData.cpp [hunk 2]
```diff
@@ -158,9 +158,136 @@ ReaderProxyData& ReaderProxyData::operator =(
...
+    }
+    if (m_qos.m_topicData.send_always() || m_qos.m_topicData.hasChanged)
+    {
+        ret_val += m_qos.m_topicData.cdr_serialized_size();
+    }
+    if (m_qos.m_groupData.send_always() || m_qos.m_groupData.hasChanged)
+    {
+        ret_val += m_qos.m_groupData.cdr_serialized_size();
+    }
+    if (m_qos.m_timeBasedFilter.send_always() || m_qos.m_timeBasedFilter.hasChanged)
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/data/ReaderProxyData.cpp]: We are adding this twice! Also in `writeToCDRMessage`!

File: src/cpp/rtps/builtin/data/ReaderProxyData.cpp [hunk 3]
```diff
@@ -158,9 +158,136 @@ ReaderProxyData& ReaderProxyData::operator =(
...
+    }
+    if (m_type_id && m_type_id->m_type_identifier._d() != 0)
+    {
+        ret_val += m_type_id->cdr_serialized_size();
+    }
+    if (m_type && m_type->m_type_object._d() != 0)
+    {
+        ret_val += m_type->cdr_serialized_size();
+    }
+    if (m_type_information && m_type_information->assigned())
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/data/ReaderProxyData.cpp]: Missing `m_qos.type_consistency`

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1512/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1284/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1510/)

**Pull request merged**